### PR TITLE
v3 glamor package with typescripting and tests 

### PR DESCRIPTION
**Pull request by mosbahh**

### CODE REVIEW

**Comment by threepointone** [comment]
Fanstastic, will dive into this today/tomorrow, lets aim to merge in by the weekend.

**Comment by otbe** [comment]
I guess there is missing a WIP label on this PR :)
We need to figure out some more things (build pipeline for TS, broken CI, 2-3 possible bugs, cleanup deps)
Hopefully done by the end of this week :)

**Comment by threepointone** [comment]
No hurry! Better slow and correct, than fast and broken

**Comment by mosbahh** [comment]
I think you can now review the PR @threepointone @donaldpipowitch :)

**Comment by threepointone** [comment]
I'll have a look. Thank you!

File: packages/glamor/src/StyleSheet.ts [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+function sheetForTag(tag: HTMLStyleElement) {
+    if (tag.sheet) {
+        return tag.sheet as CSSStyleSheet;
+    }
+
+    // this weirdness brought to you by firefox
+    // tslint:disable-next-line:prefer-for-of
+    for (let i = 0; i < document.styleSheets.length; i++) {
+        if (document.styleSheets[i].ownerNode === tag) {
```
Comment (dzearing) [inline packages/glamor/src/StyleSheet.ts]: minor nit, seems like the PR is inconsistent on spacing; 2 vs 4... (personally prefer 2 :)

**Comment by threepointone** [comment]
what say we publish a beta/RC with this?

**Comment by donaldpipowitch** [comment]
@mosbahh I started testing your PR in some of our internal projects and noticed that there are internal types which aren't re-exported, but occur in the public API.

For example we have this in our generated `index.d.ts` for `glamor` (_stripped out non-relevant code_):

```ts
import { Rule, StyleAttribute } from './css';
declare function css(...rules: Array<Rule>): StyleAttribute;
export { css };
```

AFAIK it should look something like this (thoughts @otbe?):

```ts
import { Rule, StyleAttribute } from './css';
declare function css(...rules: Array<Rule>): StyleAttribute;
export { css, Rule, StyleAttribute };
```

Now you can build abstractions on top of if in your custom code :

```
import { css, Rule, StyleAttribute } from 'glamor';

export function doSomethingCool(...rules: Array<Rule>): StyleAttribute {
  // do something with rules...
  return css(rules);
}

// if I pass something invalid to `doSomethingCool` it will tell me about it
// directly when I (or others) use `doSomethingCool`
```

**Comment by donaldpipowitch** [comment]
Thank you for your work so far 👍  ❤️

**Comment by mosbahh** [comment]
@donaldpipowitch do we need to export all the internal defined types that have been used by all exported APIs or only the ones that are related to the main css function?
for example: we exported css.global which takes a style from type "CSSProperties" as attribute, should we also export CSSProperties?!

**Comment by donaldpipowitch** [comment]
Sadly we need to export all types which occur in public APIs, not just related to `css` 😭

**Comment by threepointone** [comment]
what's up my peeps

**Comment by threepointone** [comment]
is this typescript rewrite becoming too complicated?

**Comment by donaldpipowitch** [comment]
@mosbahh Do you want to finish this PR? Do you have enough time? Or should I finish the last step?

**Comment by donaldpipowitch** [comment]
Forgotten push 😄  I'll look into the CI error.

**Comment by donaldpipowitch** [comment]
@mosbahh The error says:

```
lerna ERR! execute src/css/index.ts(113,54): error TS2559: Type 'CSSProperties' has no properties in common with type 'DeconstrucedStyles'.
lerna ERR! execute src/css/index.ts(118,54): error TS2559: Type 'CSSProperties' has no properties in common with type 'DeconstrucedStyles'.
```

It happens because a newer version of TypeScript caught a bug which the older one didn't.

AFAIK the type of [`DeconstrucedStyles `](https://github.com/mosbahh/glamor/blob/c0d8633149e4d21dad9a09d48e984a693638a3a4/packages/glamor/src/css/index.ts#L8) needs to be:

```ts
export interface DeconstrucedStyles {
  plain?: CSSProperties;
  selects?: { [key: string]: DeconstrucedStyles };
  medias?: { [key: string]: DeconstrucedStyles };
  supports?: { [key: string]: DeconstrucedStyles };
}
```

👍

**Comment by donaldpipowitch** [comment]
I just tested this again in a local project and it seems to work fine. Thanks a lot. 💃 

I just wanted to point out that some top-level aliases are "missing" (e.g. no `insertRule`, but `css.insert`; no `insertGlobal`, but `css.global`...). I don't know if we ever _explicitly_ discussed this, but I think it is total in the sense of https://github.com/threepointone/glamor/issues/83 (making `css` the bread and butter function). AFAIK they were left out on purpose. So I'm fine with this.

Without further ado: LGTM. 🎉 

If there are no objections I'll merge this in the next 24 to 48 hours or so 😄

**Comment by threepointone** [comment]
Maybe it's not worth breaking backward compat just for those. Sorry for changing my mind, but would you bring back those two aliases? 

Otherwise, feel free for to merge and publish a beta!

**Comment by donaldpipowitch** [comment]
I can do that after the merge.

These ones, too?

```
// currently missing
export const fontFace = css.fontFace
export const keyframes = css.keyframes
```

**Comment by threepointone** [comment]
Yes please

**Comment by donaldpipowitch** [comment]
Thank you all 🎉

**Comment by donaldpipowitch** [comment]
FYI: Added the missing exports in https://github.com/threepointone/glamor/commit/ee7cb3c765ec5591afd6175bbd703fa5e19b09d4.

**Comment by mosbahh** [comment]
Thank you all :+1:

**Pull request merged**